rust/src/bs_safe.rs

```diff
@@ -0,0 +1,610 @@
+                out.as_mut_ptr(),
+                rng.map(|rng| rng.rng.as_ptr()).unwrap_or(null_mut()),
+                err.as_ptr(),
+            )
+        };
+
+        if rc == 0 {
+            Ok(())
+        } else {
+            Err(BridgeStanError::EvaluationFailed(err.message()))
+        }
+    }
+
+    /// Set the sequence of unconstrained parameters based on the
+    /// specified constrained parameters, and return a return code of 0
+    /// for success and -1 for failure.  Parameter order is as declared
+    /// in the Stan program, with multivariate parameters given in
+    /// last-index-major order.
+    ///
+    /// # Arguments
+    /// `theta`: The vector of constrained parameters
+    ///
+    /// `theta_unc` Vector of unconstrained parameters
+    pub fn param_unconstrain(&self, theta: &mut [f64], theta_unc: &mut [f64]) -> Result<()> {
+        assert_eq!(
+            theta_unc.len(),
+            self.param_unc_num(),
+            "Argument 'theta_unc' must be the same size as the number of parameters!"
+        );
+        assert_eq!(
+            theta.len(),
+            self.param_num(false, false),
+            "Argument 'out' must be the same size as the number of parameters!"
+        );
+        let mut err = ErrorMsg::new(self.lib.borrow());
+        let rc = unsafe {
+            self.ffi_lib().bs_param_unconstrain(
+                self.model.as_ptr(),
+                theta.as_ptr(),
+                theta_unc.as_mut_ptr(),
+                err.as_ptr(),
+            )
+        };
+        if rc == 0 {
+            Ok(())
+        } else {
+            Err(BridgeStanError::EvaluationFailed(err.message()))
+        }
+    }
+
+    /// Map a constrained point in json format to the unconstrained space.
+    ///
+    /// The JSON schema assumed is fully defined in the *CmdStan Reference Manual*.
+    /// A value for each parameter in the Stan program should be provided, with
+    /// dimensions and size corresponding to the Stan program declarations.
+    pub fn param_unconstrain_json<S: AsRef<CStr>>(
+        &self,
+        json: S,
+        theta_unc: &mut [f64],
+    ) -> Result<()> {
+        assert_eq!(
+            theta_unc.len(),
+            self.param_unc_num(),
+            "Argument 'theta_unc' must be the same size as the number of parameters!"
+        );
+        let mut err = ErrorMsg::new(self.lib.borrow());
+        let rc = unsafe {
+            self.ffi_lib().bs_param_unconstrain_json(
+                self.model.as_ptr(),
+                json.as_ref().as_ptr(),
+                theta_unc.as_mut_ptr(),
+                err.as_ptr(),
+            )
+        };
+        if rc == 0 {
+            Ok(())
+        } else {
+            Err(BridgeStanError::EvaluationFailed(err.message()))
+        }
+    }
+}
+
+impl<T: Borrow<StanLibrary> + Clone> Model<T> {
+    /// Return a clone of the underlying stan library
+    pub fn clone_library_ref(&self) -> T {
+        self.lib.clone()
+    }
+}
```

Comment:
Is this necessary (or nice-to-have, in which case, for what)?

Reply:
Purely a nice-to-have I think. The use case I had in mind was that if `Borrow` is an `Arc` or `Rc`, users of the API can get a new reference to the library through that function, for instance to get a separate model instance with a different dataset. Without this a user would always have to keep their own reference to the library if that would ever be needed again.